docs/source/notebooks/quantum_kernel_methods.ipynb

```diff
@@ -0,0 +1,1234 @@
+    "\n",
+    "where $\\mathbf{s}! = s_1! s_2! \\cdots s_M!$. We will also consider the kernel constructed using distinguishable photons. In this case, we have the classical \"coherent kernel\":\n",
+    "$$ \\kappa_C^{(\\mathbf{s})}(x, x') = \\frac{\\text{Per} \\left( |U_{s, s} (x, x')|^{2} \\right)}{\\mathbf{s}!} $$\n",
+    "\n",
+    "There exists an efficient classical algorithm to estimate the permanent $\\text{Per} \\left( |U_{s, s} (x, x')|^{2} \\right)$ [3].\n",
+    "\n",
+    "In experimental circumstances, the experimentalist may not have access to photon-number resolving (PNR) detectors. In this instance, we consider the \"unbunching kernel\" in which there is maximum 1 photon found in each mode. In this circumstance, we lose the commutative property of the kernel function and the kernel matrix is not symmetric, positive definite for identical input datasets. We will force the kernel matrix to be symmetric by projecting the upper triangular onto the lower triangular. We will also force this matrix to be positive definite by decomposing the matrix spectrally:\n",
+    "$$K = Q \\Lambda Q^{T}$$\n",
+    "and setting the negative eigenvalues in $\\Lambda$ to $0$. The kernel matrix is then reconstructed."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def kernel(\n",
+    "    x1, x2: np.ndarray, \n",
+    "    feature_map: Callable[[np.ndarray], Circuit], \n",
+    "    input_state: BasicState, \n",
+    "    indistinguishability: float = 1.0, \n",
+    "    unbunching: bool = False, \n",
+    "    nshots: int = None\n",
+    ") -> float:\n",
+    "    \"\"\"\n",
+    "    Estimate the kernel associated with a given feature map.\n",
+    "    Args:\n",
+    "        x1, x2               : Data inputs\n",
+    "        feature_map          : Quantum circuit to perform feature map\n",
+    "        input_state          : Initial state quantum feature map is applied to.\n",
+    "        indistinguishability : Photon indistinguishability (Hong Ou Mandel)\n",
+    "        unbunching           : Specifies whether to not use PNR detectors\n",
+    "        nshots               : Number of circuit runs to estimate kernel\n",
+    "    Returns\n",
+    "        Scalar kernel value\n",
+    "    \"\"\"\n",
+    "    # Compile circuit to estimate kernel\n",
+    "    U = feature_map(x1)\n",
+    "    U_dag = feature_map(x2)\n",
+    "    U_dag.inverse(h=True)\n",
+    "    processor = Processor(\"SLOS\", U.m)\n",
+    "    processor.add(0, U)\n",
+    "    processor.add(0, U_dag)\n",
+    "    processor.noise = NoiseModel(indistinguishability=indistinguishability)\n",
+    "    \n",
+    "    if unbunching:\n",
+    "        processor.thresholded_output(True)\n",
+    "    \n",
+    "    # Apply settings to processor\n",
+    "    processor.with_input(input_state)\n",
```

Comment:
Can you call `processor.min_detected_photons_filter(input_state.n)` before this line (you removed the two lines where you should have kept one)